### DISCUSSION TITLE
Creating a table at runtime using predefined entity under any PSQL schema

### DISCUSSION THREAD

**Discussion by tukusejssirs** [mikro-orm/mikro-orm#2473] [Q&A]
Related: #2439

---

I need to create same table under different schemas at runtime.

We use schemas as folders to group same tables for different machines/devices. I create an entity with `@Entity({schema: '*', tableName: 'status'})`, but how can I use that entity to create a table under a particular schema (specified as a parameter)?

I also want to create that partcular schema at runtime.

Do I need to use `em.execute()`, as you suggested in [here](https://github.com/mikro-orm/mikro-orm/discussions/2237#discussioncomment-1401640)?

**Comment by B4nan** [comment]
All schema generator methods now have `schema` parameter. You can use it programatically.

https://mikro-orm.io/docs/next/multiple-schemas/#wildcard-schema
https://mikro-orm.io/docs/next/schema-generator/#using-schemagenerator-programmatically

**Reply by tukusejssirs** [reply]
Well, let’s assume I use `mikro-orm.config.ts` and a service with the following constructor:

```ts
constructor(
	@InjectRepository(Status)
	private readonly statusRepository: EntityRepository<Status>,
	private readonly orm: MikroORM,
	private readonly configService: ConfigService
) {}
```

Within that service, how can I create a new table based on the `Status` entity, which has `@Entity({schema: '*', tableName: 'status'})` decorator? How can I specify the schema name (e.g. `m1`) to create a table based on that entity (e.g. `m1.status`)?

In the [docs](https://mikro-orm.io/docs/next/schema-generator/#using-schemagenerator-programmatically), you have the following example:

```ts
const orm = await MikroORM.init({
  entities: [Author, Book, ...],
  dbName: 'your-db-name',
  // ...
});
```

I presume I can add `schema` property to the MikroORM config, but in Nest.js I use a global MikroORM configuration. Where can I set the schema name then?

**Reply by B4nan** [reply]
Well, this feels like you keep asking things without actually reading my replies. So once again, you use schema generator programatically and you pass the `schema` option there.

**Reply by tukusejssirs** [reply]
> **_All schema generator methods_** now have `schema` parameter. You can use it programatically.

Okay, sorry. I think I missed _generator methods_ when I read your comment.

As I read it in my text editor, `getCreateSchemaSQL()` method accepts one optional parameter `options` which is an object that can contain two optional properties `wrap` and `schema`. Therefore I can use `getCreateSchemaSQL({schema: 'm1'})`, right?

**Reply by tukusejssirs** [reply]
A subquestion: after creating the table, how can I access it? Do I need to update the entity consructor or something?

**Reply by B4nan** [reply]
you would edit the entity definition as usual, and use schema generator `updateSchema()` method on each schema you have (so again with the `schema` option as with `createSchema()`)

**Reply by tukusejssirs** [reply]
And when I want to build a query (e.g. `const qb = this.statusRepository.createQueryBuilder('s').select('*')`), where can I specify the schema name?

I suspect it is using `.withSchema(schemaName)` (based on [this](https://mikro-orm.io/docs/next/multiple-schemas/#wildcard-schema)). **Yes, this works indeed, sorry for the noise.**

**Reply by tukusejssirs** [reply]
Until now (when using status `schema`, not `*`), I inserted new records (rows) using an entity contructor (e.g. `new Status(values)`), which currently does not support `withSchema(schemaName)`.

How can I define the schema name when inserting new records?

I wanted to insert a new record using the following code ([src](https://mikro-orm.io/docs/next/multiple-schemas)), but I have no idea how can I define `em`. I tried to use `const em = this.orm.em`, but it throws an error that `Property 'createQueryBuilder' does not exist on type 'EntityManager<IDatabaseDriver<Connection>>'.`

```ts
const qb = em.createQueryBuilder(User);
await qb.insert({ email: 'foo@bar.com' }).withSchema('client-123');
```

I also tried to `grep` the example repo, but it uses the constructor and does not define `em` in its codebase.

**Reply by B4nan** [reply]
You can do this:

```ts
const status = new Status(...);
wrap(status).setSchema('...');
await em.persist(status).flush();
```

> I wanted to insert a new record using the following code (src), but I have no idea how can I define em. I tried to use const em = this.orm.em, but it throws an error that Property 'createQueryBuilder' does not exist on type 'EntityManager<IDatabaseDriver<Connection>>'.

You need to have the `MikroORM` instance typed to `MikroORM<PostgreSqlDriver>` (or any other SQL driver) to access methods like `createQueryBuilder`. Either use the argument in `init` method or just type cast.

You can also just cast the em instance, for that you can use `import { EntityManager } from '@mikro-orm/postgresql';` which will have the method (it is an alias for [SqlEntityManager](https://github.com/mikro-orm/mikro-orm/blob/master/packages/knex/src/SqlEntityManager.ts))

**Reply by tukusejssirs** [reply]
I see an advantage of using `await qb.insert(createStatusDto).withSchema(schema)` instead of `const status = new Status(...)`: it automatically writes to the DB without running `await this.statusRepository.persistAndFlush(status)` manually! :wink:
